tracer/src/Datadog.Trace/IAST/Analyzers/HardcodedSecretsAnalyzer.cs

```diff
@@ -0,0 +1,214 @@
+    }
+
+    private static void PoolingThread()
+    {
+        try
+        {
+            Log.Debug("HardcodedSecretsAnalyzer polling thread -> Started");
+            while (_started)
+            {
+                var userStrings = new UserStringInterop[UserStringsArraySize];
+                int userStringLen = NativeMethods.GetUserStrings(userStrings.Length, userStrings);
+                Log.Debug("HardcodedSecretsAnalyzer polling thread -> Retrieved {UserStringLen} strings", userStringLen.ToString());
+                if (userStringLen > 0 && Tracer.Instance.Settings.IsIntegrationEnabled(IntegrationId.HardcodedSecret))
+                {
+                    for (int x = 0; x < userStringLen; x++)
+                    {
+                        var value = Marshal.PtrToStringUni(userStrings[x].Value);
+                        var match = CheckSecret(value!);
+                        if (!string.IsNullOrEmpty(match))
+                        {
+                            var location = Marshal.PtrToStringUni(userStrings[x].Location);
+                            IastModule.OnHardcodedSecret(new Vulnerability(
+                                VulnerabilityTypeName.HardcodedSecret,
+                                (VulnerabilityTypeName.HardcodedSecret + ":" + location!).GetStaticHashCode(),
+                                new Location(location!),
+                                new Evidence(match!),
+                                IntegrationId.HardcodedSecret));
+                        }
+                    }
+
+                    if (userStringLen == userStrings.Length)
+                    {
+                        continue; // Skip wait time if array came full
+                    }
+                }
+
+                _waitEvent.Wait(5_000);
+            }
+        }
+        catch (Exception err)
+        {
+            _started = false;
+            Log.Warning(err, "Exception in HardcodedSecretsAnalyzer polling thread. Disabling feature.");
+        }
+
+        Log.Debug("HardcodedSecretsAnalyzer polling thread -> Exit");
+    }
+
+    internal static string? CheckSecret(string secret)
+    {
+        if (_secretRules == null)
+        {
+            _secretRules = GenerateSecretRules();
+        }
+
+        foreach (var rule in _secretRules)
+        {
+            if (rule.Regex.IsMatch(secret))
+            {
+                return rule.Rule;
+            }
+        }
+
+        return null;
+    }
+
+    internal static void Initialize()
+    {
+        lock (Log)
+        {
+            if (_instance == null)
+            {
+                _instance = new HardcodedSecretsAnalyzer();
+            }
+        }
+    }
+
+    private static List<SecretRegex> GenerateSecretRules()
+    {
+        var res = new List<SecretRegex>();
+
+        res.Add(new SecretRegex("aws-access-token", @"\b((A3T[A-Z0-9]|AKIA|AGPA|AIDA|AROA|AIPA|ANPA|ANVA|ASIA)[A-Z0-9]{16})(?:['|""|\n|\r|\s|\x60|;]|$)"));
```

Comment:
Just to double check... these Regexes don't appear to be anchored to the start or end of the string. Is that correct? We want to look _inside_ all strings, not just confirm the string provided matches exactly the pattern? 

It's fine if so, it's just that will be a lot slower and have many more potential DOS scenarios, such that we will almost certainly want to add a timeout (and use non-backtracking if possible)

Reply:
The ones meant to be anchored already are, but most of them don't. I'm adding the timeout (I had the idea I did it somewhere in time...)